src/restic/filter/filter_test.go

```diff
@@ -124,6 +124,77 @@ func TestMatch(t *testing.T) {
 	}
 }
 
+var childMatchTests = []struct {
+	pattern string
+	path    string
+	match   bool
+}{
+	{"", "", true},
+	{"", "/foo", true},
+	{"", "/x/y/z/foo", true},
+	{"foo/bar", "/foo", true},
+	{"baz/bar", "/foo", true},
+	{"foo", "/foo/bar", true},
+	{"bar", "/foo", true},
+	{"baz", "/foo/bar", true},
+	{"*", "/foo", true},
+	{"*", "/foo/bar", true},
+	{"/foo/bar", "/foo", true},
+	{"/foo/bar/baz", "/foo", true},
+	{"/foo/bar/baz", "/foo/bar", true},
+	{"/foo/bar/baz", "/foo/baz", false},
+	{"/foo/**/baz", "/foo/bar/baz", true},
+	{"/foo/**/qux", "/foo/bar/baz/qux", true},
+	{"/baz/bar", "/foo", false},
+	{"/foo", "/foo/bar", true}, // maybe this should be false
+	{"/*", "/foo", true},       // maybe this should be false
+	{"/*", "/foo/bar", true},   // maybe this should be false
+	{"/foo", "/foo/bar", true}, // maybe this should be false
```

Comment:
There may be some semantic choice to be made here: `/foo` and `/foo/*` are equivalent to `/foo/**` currently. Maybe a single `*` could mean "restore just files at that level and `foo` or `/foo` mean "restore `foo` but if it's a dir, don't go down". I kept the existing behavior as is.